pyqtgraph/widgets/_plotitemoverlay.py

```diff
@@ -0,0 +1,256 @@
+
+    def add_plotitem(
+        self,
+        plotitem: PlotItem,
+
+        # TODO: we could also put the ``ViewBox.XAxis``
+        # style enum here?
+        link_axes: tuple[int] = (),
+
+        # XXX: all the other optional inputs.
+        # should we enum or at least type check this?
+        # link_axes: tuple[int] = (0,),  # link x
+        # link_axes: tuple[int] = (1,),  # link y
+        # link_axes: tuple[int] = (0, 1),  # link both
+
+    ) -> None:
+
+        root = self.root_plotitem
+        layout: QGraphicsGridLayout = root.layout
+        self.overlays.append(plotitem)
+
+        vb: ViewBox = plotitem.vb
+
+        # can't do this since only root will get
+        # menu event then..
+        # TODO: some sane way to allow menu event broadcast XD
+        # vb.setMenuEnabled(False)
+
+        # instruct all vbs to relay signals instead
+        # of consuming without relay
+        vb.allow_signal_relay = True
+
+        # TODO: move this into ``ViewBox`` as some kind of special
+        # linking method for full view box event relaying.
+        if not link_axes:
+            # TODO: there's still an issue with manually click-dragging
+            # axes - not sure what that's about but likely some bug
+            # inside the mangled mess of event handling that is the
+            # `ViewBox` core XD
+
+            # NOTE: for this to work the ``.allow_signal_relay`` patch
+            # made to ``ViewBox`` needs to exist to avoid events being
+            # consumed too early.
+            self._relays[plotitem] = root.vb.sigMouseDragged.connect(
+                vb.mouseDragEvent)
+            self._relays[plotitem] = root.vb.sigMouseWheel.connect(vb.wheelEvent)
+
+        else:
+            for dim in link_axes:
+                # link x and y axes to new view box such that the top level
+                # viewbox propagates to the root (and whatever other plotitem
+                # overlays that have been added).
+                # vb.linkView(dim, root.vb)
+                root.vb.linkView(dim, vb)
+
+        # make overlaid viewbox impossible to focus since the top
+        # level should handle all input and relay to overlays.
+        # TODO: we will probably want to add a "focus" api such that
+        # a new "top level" ``PlotItem`` can be selected dynamically
+        # (and presumably the axes dynamically sorted to match).
+        vb.setFlag(
+            vb.GraphicsItemFlag.ItemIsFocusable,
+            False
+        )
+        vb.setFocusPolicy(Qt.NoFocus)
+
+        # Add any axes in appropriate sequence to the top level layout
+        # to avoid graphics collision.
+        count = len(self.overlays)
+        assert count
+
+        for name, axis_info in plotitem.axes.items():
+            axis = axis_info['item']
+            axis_view = axis.linkedView()
+
+            # Remove old axis
+            # plotitem.removeAxis(axis, unlink=False)
+            if not axis.isVisible():
```

Comment:
We should be able to drop this if we make `PlotItem.setAxisItems()` actually be a pure write operation.